crates/apollo-compiler/src/validation/selection.rs

```diff
@@ -1,6 +1,273 @@
+                        original_selection: field_a.loc().into(),
+                        redefined_selection: field_b.loc().into(),
+                    },
+                )
+                .label(Label::new(arg.loc(), format!("field `{}` is selected with argument `{}` here", field_b.name(), arg.name())))
+                .label(Label::new(field_a.loc(), format!("but argument `{}` is not provided here", arg.name())))
+                .help("Fields with the same response name must provide the same set of arguments. Consider adding an alias if you need to select fields with different arguments.")
+            );
+        };
+    }
+
+    Ok(())
+}
+
+/// Check if the fields in a given selection set can be merged.
+///
+/// Spec: https://spec.graphql.org/October2021/#FieldsInSetCanMerge()
+fn fields_in_set_can_merge(
+    db: &dyn ValidationDatabase,
+    selection_set: &hir::SelectionSet,
+) -> Result<(), ApolloDiagnostic> {
+    // 1. Let `fieldsForName` be the set of selections with a given response name in set including visiting fragments and inline fragments.
+    let fields = db.flattened_operation_fields(selection_set.clone());
+    let grouped_by_name = group_fields_by_name(fields);
+
+    for (_, fields_for_name) in grouped_by_name {
+        // 2. Given each pair of members fieldA and fieldB in fieldsForName:
+        let Some((field_a, rest)) = fields_for_name.split_first() else {
+            continue; // Nothing to merge
+        };
+        let Some(parent_a) = field_a.parent_type(db.upcast()) else {
+            continue; // We can't find the type
+        };
+        for field_b in rest {
+            // 2a. SameResponseShape(fieldA, fieldB) must be true.
+            same_response_shape(db, field_a, field_b)?;
+            // 2b. If the parent types of fieldA and fieldB are equal or if either is not an Object Type:
+            let Some(parent_b) = field_b.parent_type(db.upcast()) else {
+                continue; // We can't find the type
+            };
+            if parent_a == parent_b {
+                // 2bi. fieldA and fieldB must have identical field names.
+                if field_a.name() != field_b.name() {
+                    return Err(ApolloDiagnostic::new(
+                        db,
+                        field_b.loc().into(),
+                        DiagnosticData::ConflictingSelection {
+                            field: field_a.name().to_string(),
+                            original_selection: field_a.loc().into(),
+                            redefined_selection: field_b.loc().into(),
+                        },
+                    )
+                    .label(Label::new(
+                        field_a.loc(),
+                        format!(
+                            "field `{}` is selected from field `{}`",
+                            field_a.response_name(),
+                            field_a.name()
+                        ),
+                    ))
```

Comment:
was thinking some more about this. 

i don't think i follow the diagnostic for the first screenshot titled "Conflicting fields aliased to the same name".

the validation issue is that we have an alias that's bares the same name as another field, and the two of them return different types (and therefore cannot be merged. what if we say something along the lines of "this alias {alias_name} references field {field_name} and returns type {type_definition_name}" for the original aliased reference, and for the field name definition something like this: "{alias_name} is redefined here and returns type {type_definition_name}". The `Note` portion of the diagnostic can then describe the validation rule in a bit more detail.

(field merging is a complicated rule, we are going to have to iterate on these diagnostics a few times, i am sure)

Reply:
I read this particular rule ([`FieldsInSetCanMerge` 2.b.i.](https://spec.graphql.org/October2021/#FieldsInSetCanMerge())) differently: it specifically checks that two fields that have the same response name (i.e. alias if configured, else the field name) have the same field name. It's for cases like this:
```graphql
query GetUser {
  name: displayName
  name: legalName
}
```
Even if `displayName` and `legalName` have the same types, this is not allowed.

This rule has to exist because it *is* allowed to do
```graphql
query GetUser {
  name
  name
}
```
which just results in the `name` field being selected with response name `name`.

But if a response name `name` is selected twice *from different fields*, that's a conflict.

Differing types are checked by `SameResponseShape`, so that's a separate rule

Reply:
so we do need to improve the diagnostic message to clarify the above :D

Reply:
ahhhhh! okay, yea let's think of a clearer diagnostic messaging then 🤔 🤔